### ISSUE TITLE
[Question]: Exposing common Playwright types

### ISSUE THREAD

**Issue by smhutch** [microsoft/playwright#4255]
## TL;DR

Would it be possible to export common types like `HTMLOrSVGElementHandle` and `PageFunction` from the playwright core types?

## Details

I have been using Playwright to implement integration tests, some of which get quite long (we're testing a lot). To make these tests a bit more maintainable, I have started to abstract certain testing logic that we re-use in many places. For example, here's an example of a block of code that I abstracted today:

```ts
// Doesn't work, not exported currently.
import { HTMLOrSVGElementHandle } from 'playwright';

import { selectableSelectors } from './selectors';

type CheckOptions = Record<'deselectedCount' | 'selectedCount', number>;
type CheckHandlers = Record<
  'deselectedElements' | 'selectedElements',
  HTMLOrSVGElementHandle[]
>;

export const checkSelectableItems = async (
  options: CheckOptions
): CheckHandlers => {
  // Get selected/deselected elements in selectable
  const selectedElements = await page.$$(selectableSelectors.selected);
  const deselectedElements = await page.$$(selectableSelectors.deselected);

  // Check that number of checked items matches expected
  expect(selectedElements).toHaveLength(options.selectedCount);
  expect(deselectedElements).toHaveLength(options.deselectedCount);

  // Return elements for further testing
  return {
    deselectedElements,
    selectedElements,
  };
};
```

This abstraction does what I intended it to do, but will fail type checking because `HTMLOrSVGElementHandle` is not exported from playwright. Since my project enforces explicit return types on module boundaries, I have a few options:

1. Disable the lint rule, and allow TS to infer the return type. This is the best option I have right now.
2. Write very verbose TypeScript to "extract" the return type from `page.$$`, for example:
```ts
type PlayrightHandlesPromise = ReturnType<typeof page.$$>;
type ThenCallback = Parameters<PlayrightHandlesPromise['then']>[0];
// This gives me the desired type
type ForcePlayrightHandle = Parameters<ThenCallback>[0];
```
3. Add an unsafe return type like `any`, which "fixes" my lint error, but results in far less type-safety in tests which consume this function.

**Comment by JoelEinbinder** [comment]
`HTMLOrSVGElementHandle` is just our internal shortcut for `ElementHandle<HTMLElement|SVGElement>`. I agree that we shouldn't use types that we don't export. However I'm thinking of just removing `HTMLOrSVGElementHandle` rather than exporting the shortcut.

Would that work or are you specifically after the shortcut?

Page function is a different story. I can see how that would be helpful for creating wrapper functions. Do you have an example for this? I'm thinking about whether you just need `PageFunction` or also need `Unboxed` and `SmartHandle`.

**Comment by smhutch** [comment]
@JoelEinbinder Thanks! Although accessing the shortcut would make types at my end marginally simpler, using `ElementHandle<HTMLElement|SVGElement>` is good enough to solve my problem for now 🙂.

I ended up with:

```ts
import { ElementHandle } from 'playwright';

type CheckReturn = Record<
  'deselectedElements' | 'selectedElements',
  ElementHandle<HTMLElement | SVGElement>[]
>;

export const checkSelectableItems = async (
  options: CheckOptions
): Promise<CheckReturn> => { ... }
```

Regarding `PageFunction`, I don't have an example of this right now. I chose it purely because it seems like one I **might** use in the future. If you feel that others may benefit from some additional exported types it might be worth exporting that. Otherwise, feel free to close this, you answered my current question 👍

**Comment by arjunattam** [comment]
Thanks for confirming @smhutch! I'll close this issue for now. We can re-open if there's interest in using `PageFunction`.
